seahorn/lib/linked_list_helper.c

```diff
@@ -64,89 +27,259 @@ static inline void init_short_aws_linked_list(struct aws_linked_list *list,
+    ;
+  } else if (saved->saved_size == 2) {
+    // if saved size is 2 then check start, start.next and tail are unchanged
+    return nodes_next_equal(saved->save_point, &saved->nodes[0]) &&
+           nodes_equal(saved->save_point->next, &saved->nodes[1]) &&
+           nodes_equal(&list->tail, &saved->tail);
+    ;
+  } else if (saved->saved_size == 3) {
+    // we only save a maximum of three nodes since that is the maximum we
+    // construct concretely are head <--> node1 <--> node2
+    return nodes_next_equal(saved->save_point, &saved->nodes[0]) &&
+           nodes_equal(saved->save_point->next, &saved->nodes[1]) &&
+           nodes_equal(saved->save_point->next->next, &saved->nodes[2]) &&
+           nodes_equal(&list->tail, &saved->tail);
   } else {
-    // HEAD <--> front --> nd ... nd <-- back <--> TAIL
-    struct aws_linked_list_node *front = malloc(sizeof(struct aws_linked_list_node));
-    struct aws_linked_list_node *back = malloc(sizeof(struct aws_linked_list_node));
-    aws_linked_list_attach_after(&list->head, front, true);
-    aws_linked_list_attach_after(front, back, false);
-    aws_linked_list_attach_after(back, &list->tail, true);
+    return false;
   }
+  // do nothing here: if condition should be exhaustive
+}
+
+bool is_aws_list_unchanged_to_head(struct aws_linked_list *list,
+                                   struct saved_aws_linked_list *saved) {
+  if (saved->saved_size == 0) {
+    // if saved size is zero, only check that head is unchanged
+    return nodes_prev_equal(&list->head, &saved->head);
+  } else if (saved->saved_size == 1) {
+    // if saved size is 1 then check start and head are unchanged
+    return nodes_prev_equal(saved->save_point, &saved->nodes[0]) &&
+           nodes_equal(&list->head, &saved->head);
+    ;
+  } else if (saved->saved_size == 2) {
+    // if saved size is 2 then check start, start.prev and head are unchanged
+    return nodes_prev_equal(saved->save_point, &saved->nodes[0]) &&
+           nodes_equal(saved->save_point->prev, &saved->nodes[1]) &&
+           nodes_equal(&list->head, &saved->head);
+    ;
+  } else if (saved->saved_size == 3) {
+    // we only save a maximum of three nodes since that is the maximum we
+    // construct concretely are node1 <--> node2 <--> tail
+    return nodes_prev_equal(saved->save_point, &saved->nodes[0]) &&
+           nodes_equal(saved->save_point->prev, &saved->nodes[1]) &&
+           nodes_equal(saved->save_point->prev->prev, &saved->nodes[2]) &&
+           nodes_equal(&list->head, &saved->head);
+  } else {
+    return false;
+  }
+  // do nothing here: if condition should be exhaustive
+}
+
+// iterator function for save*Node
+struct aws_linked_list_node *getNext(struct aws_linked_list_node *node) {
+  return node->next;
+}
+
+// iterator function for save*Node
+struct aws_linked_list_node *getPrev(struct aws_linked_list_node *node) {
+  return node->prev;
+}
+
+void saveOneNode(struct aws_linked_list_node *start,
```

Comment:
would be nice to follow the same naming convention. Since this PR already has a show stopper, let's change this at the same time.

Reply:
done